server/src/main/java/org/elasticsearch/cluster/ClusterInfo.java

```diff
@@ -224,7 +224,36 @@ public ReservedSpace getReservedSpace(String nodeId, String dataPath) {
      * includes a 'p' or 'r' depending on whether the shard is a primary.
      */
     public static String shardIdentifierFromRouting(ShardRouting shardRouting) {
-        return shardRouting.shardId().toString() + "[" + (shardRouting.primary() ? "p" : "r") + "]";
+        return shardIdentifierFromRouting(shardRouting.shardId(), shardRouting.primary());
+    }
+
+    public static String shardIdentifierFromRouting(ShardId shardId, boolean primary) {
+        return shardId.toString() + "[" + (primary ? "p" : "r") + "]";
+    }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
We should use `EqualsHashCodeTestUtils.checkEqualsAndHashCode` to test this in `ClusterInfoTests`. Alternatively add a dedicated `AbstractWireSerializingTestCase` derivative for `ClusterInfo` which would cover equals and hashcode as well as replacing `ClusterInfoTests#testSerialization`.

Reply:
Should the change to equals/hashCode and their tests be ported to main separately?